src/cc/BPF.h

```diff
@@ -0,0 +1,153 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cctype>
+#include <memory>
+#include <string>
+
+#include "BPFTable.h"
+#include "bcc_syms.h"
+#include "bpf_module.h"
+#include "common.h"
+#include "compat/linux/bpf.h"
+#include "libbpf.h"
+
+namespace ebpf {
+
+enum class bpf_attach_type {
+  probe_entry,
+  probe_return
+};
+
+struct open_probe_t {
+  void* reader_ptr;
+  std::string func;
+};
+
+class BPF {
+public:
+  static const int BPF_MAX_STACK_DEPTH = 127;
+
+  explicit BPF(const std::string& bpf_program, unsigned int debug_flag = 0,
+               std::vector<std::string> cflags = {});
+  StatusTuple init();
+
+  ~BPF();
+  StatusTuple detach_all();
+
+  StatusTuple attach_kprobe(
+      const std::string& kernel_func, const std::string& probe_func,
+      bpf_attach_type attach_type = bpf_attach_type::probe_entry,
+      pid_t pid = -1, int cpu = 0, int group_fd = -1,
+      perf_reader_cb cb = nullptr, void* cb_cookie = nullptr);
+  StatusTuple detach_kprobe(
+      const std::string& kernel_func,
+      bpf_attach_type attach_type = bpf_attach_type::probe_entry);
+
+  StatusTuple attach_uprobe(
+      const std::string& binary_path, const std::string& symbol,
+      const std::string& probe_func, uint64_t symbol_addr = 0,
+      bpf_attach_type attach_type = bpf_attach_type::probe_entry,
+      pid_t pid = -1, int cpu = 0, int group_fd = -1,
+      perf_reader_cb cb = nullptr, void* cb_cookie = nullptr);
+  StatusTuple detach_uprobe(
+      const std::string& binary_path, const std::string& symbol,
+      uint64_t symbol_addr = 0,
+      bpf_attach_type attach_type = bpf_attach_type::probe_entry);
+
+  StatusTuple attach_tracepoint(const std::string& tracepoint,
+                                const std::string& probe_func,
+                                pid_t pid = -1, int cpu = 0, int group_fd = -1,
+                                perf_reader_cb cb = nullptr,
+                                void* cb_cookie = nullptr);
+  StatusTuple detach_tracepoint(const std::string& tracepoint);
+
+  template <class KeyType, class ValueType>
+  BPFHashTable<KeyType, ValueType> get_hash_table(const std::string& name) {
+    return BPFHashTable<KeyType, ValueType>(bpf_module_.get(), name);
+  }
+
+  BPFStackTable get_stack_table(const std::string& name) {
+    return BPFStackTable(bpf_module_.get(), name);
+  }
+
+  StatusTuple open_perf_buffer(const std::string& name, perf_reader_raw_cb cb,
```

Comment:
I'm not a fan of the inband signaling with cpu = -1. Especially when further down close(-1) is used.
What is the use case of specifying a cpu here? Can open_on_cpu() be private and we always create ring buffer on all cpus by default? If later some use case pops up that needs exact cpu number, we can add another explicit method for that.